### PR TITLE
Function to aggregate IP prefixes

### PR DESCRIPTION
**Pull request by matt-calder**
## Description
A function `ip_prefix_collapse` which accepts an ARRAY of `IPPREFIX`s, collapses them into their minimal representation, and returns an `ARRAY` of new `IPPREFIX`s. Detailed background of problem in [https://github.com/prestodb/presto/issues/23375](https://github.com/prestodb/presto/issues/23375). 

## Motivation and Context
Addresses [https://github.com/prestodb/presto/issues/23375](https://github.com/prestodb/presto/issues/23375)

## Impact
This introduces a new user-facing SQL function called `ip_prefix_collapse`.

## Test Plan
unit tests

`mvn clean install -Dtest=TestIpPrefixFunctions -Dmaven.javadoc.skip=true -DskipUI -T1C -fn -pl presto-main`

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
*  Add function :func:`ip_prefix_collapse` ... :pr:`23445`
```

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/ip.rst
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/ip.rst:56]:
```suggestion
    Every ``IPPREFIX`` in the input array must be the same IP version (that is, only IPv4 or only IPv6). ::
```
* Adding an `s` after the ``IPPREFIX`` here causes the following error in a local doc build: 

  `/Users/steveburnett/Documents/GitHub/presto/presto-docs/src/main/sphinx/functions/ip.rst:55: WARNING: Inline literal start-string without end-string.` 
  Suggest rephrase to avoid the warning. 

* Do not use Latin abbreviations. See the [GitLab documentation style guide recommended word list entry for i.e.](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#ie).

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the doc! Just two nits, with suggested changes.

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs) 

Pull updated branch, new local doc build, docs look good. Thanks!

**Review by elharo — COMMENTED** [review]
This seems quite specific. Are there functions like this in any other SQL implementation? Can this be an extension function instead of a core function?

File: presto-docs/src/main/sphinx/functions/ip.rst
Comment (elharo) [inline presto-docs/src/main/sphinx/functions/ip.rst:56]: what if that isn't true? What happens?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (elharo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java:52]: private

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (elharo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java:202]: Isn't this guaranteed by sorting above?

**Comment by matt-calder** [comment]
Thank you for the feedback.

> This seems quite specific. 

This function is commonly found in libraries for manipulating networking data and serves a real need big data networking analytics. The most popular one in the analytics work is probably Python's `ipaddress.collapse_addresses` function. However, there are a number of implementations:
* [ip range](https://manpages.org/iprange)
* [aggregate6](https://pypi.org/project/aggregate6/)
* [netaddr.cidr_merge](https://netaddr.readthedocs.io/en/latest/api.html?highlight=merge#netaddr.cidr_merge)
* [https://pypi.org/project/aggregate-prefixes/](https://pypi.org/project/aggregate-prefixes/)
* [caprica](https://github.com/google/capirca/blob/master/capirca/lib/summarizer.py)

> Are there functions like this in any other SQL implementation? 

Yes, PostgreSQL's [inet_merge](https://pgpedia.info/i/inet_merge.html) is similar.

> Can this be an extension function instead of a core function?

* I think since there [are existing `IPPREFIX` functions](https://prestodb.io/docs/current/functions/ip.html) (e.g., `is_subnet_of`, `ip_subnet_range`), it makes sense to keep them together.
* The `IPADDRESS` and `IPPREFIX` types and supporting functions are not in presto-common, only presto-main, which prevents writing extension functions for them (classes in presto-main are on the classpath as test-only).

File: presto-docs/src/main/sphinx/functions/ip.rst
Comment (matt-calder) [reply presto-docs/src/main/sphinx/functions/ip.rst]: Updated doc.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (matt-calder) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java]: I tried to rephrase the description that the set is determined by the CIDR representation of IP ranges.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (matt-calder) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java]: Right. I've removed this condition.

**Comment by elharo** [comment]
Yes, but Presto is not a library for manipulating networking data. It's a SQL database (or execution engine, or analyzer, or whatever). There are many things commonly found in non-core libraries for Java/Python/Rust/etc. that aren't right for a SQL library. We would not, for example, want to include equivalents of everything in the standard Java class libraries, much less everything in Apache Commons, Guava, Codehaus, and many others. 

The more I think about this the more I'm convinced this should be a UDF and not something we ship in core. If you make it a UDF, you can use it tomorrow for whatever you need without waiting for any approval here.

**Comment by matt-calder** [comment]
> Yes, but Presto is not a library for manipulating networking data. It's a SQL database (or execution engine, or analyzer, or whatever). There are many things commonly found in non-core libraries for Java/Python/Rust/etc. that aren't right for a SQL library. We would not, for example, want to include equivalents of everything in the standard Java class libraries, much less everything in Apache Commons, Guava, Codehaus, and many others.

It isn't manipulating, it is analyzing. And Presto and all other major SQL engines support functions for analyzing networking data. 

> 
> The more I think about this the more I'm convinced this should be a UDF and not something we ship in core. If you make it a UDF, you can use it tomorrow for whatever you need without waiting for any approval here.

It cannot be made a UDF for the reason I mentioned previously -- The IPADDRESS and IPPREFIX types and supporting functions are not in presto-common, only presto-main, which prevents writing extension functions for them (classes in presto-main are on the classpath as test-only).

**Comment by kaikalur** [comment]
@NikhilCollooru  this is a useful function. Let's review it and merge it ASAP

**Comment by kaikalur** [comment]
> > Yes, but Presto is not a library for manipulating networking data. It's a SQL database (or execution engine, or analyzer, or whatever). There are many things commonly found in non-core libraries for Java/Python/Rust/etc. that aren't right for a SQL library. We would not, for example, want to include equivalents of everything in the standard Java class libraries, much less everything in Apache Commons, Guava, Codehaus, and many others.
> 
> It isn't manipulating, it is analyzing. And Presto and all other major SQL engines support functions for analyzing networking data.
> 
> > The more I think about this the more I'm convinced this should be a UDF and not something we ship in core. If you make it a UDF, you can use it tomorrow for whatever you need without waiting for any approval here.
> 
> It cannot be made a UDF for the reason I mentioned previously -- The IPADDRESS and IPPREFIX types and supporting functions are not in presto-common, only presto-main, which prevents writing extension functions for them (classes in presto-main are on the classpath as test-only).

Given we have IPAddress as a built-in type - this is quite reasonable ot have it as part of built-in.

**Comment by kaikalur** [comment]
Also this subtle distinction between udf vs builtin is irrelevant for this PR. If/when there is a clear definition we can refactor it. Let's unblock this here for now.

**Comment by elharo** [comment]
Is it feasible to move the IPADDRESS and IPPREFIX types to presto-common? I.e. is it a simple refactor or does it have other implications and effects?

File: presto-docs/src/main/sphinx/functions/ip.rst
Comment (elharo) [inline presto-docs/src/main/sphinx/functions/ip.rst:57]: Is there a reason for the double colon here or is that a typo?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (elharo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java:147]: Is this smallest or fewest? That is, is it returning a single smallest range that covers everything (smallest range)  or a list of ranges that covers everything (fewest ranges)?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (elharo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java:210]: use a longer try block to avoid splitting declaration and assignment. All code that depends on asInetAddress being set can go into the try block.

File: presto-docs/src/main/sphinx/functions/ip.rst
Comment (matt-calder) [reply presto-docs/src/main/sphinx/functions/ip.rst]: fixed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (matt-calder) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java]: Agreed -- "fewest ranges" is better.

**Comment by matt-calder** [comment]
> Is it feasible to move the IPADDRESS and IPPREFIX types to presto-common? I.e. is it a simple refactor or does it have other implications and effects?

I briefly looked into it but it was outside my depth for how all the presto projects fit together.

**Comment by tdcmeehan** [comment]
> Given we have IPAddress as a built-in type - this is quite reasonable ot have it as part of built-in.

Agreed.  We already support the IPADDRESS type, and this seems like a reasonable thing to do with IP addresses.  Also, for reference, Postgres supports this type built in as well.

The naming might be confusing.  AFAIA `_agg` functions alway represent aggregation functions, but technically this is a scalar function.  I wonder if something like `ip_prefix_collapse` or `ip_prefix_union` is more appropriate?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java:148]: Can you rename to get rid of "agg" to avoid confusion?
Overall lgtm. Thanks for the contribution.

**Review by feilong-liu — APPROVED** [review]

**Review by feilong-liu — APPROVED** [review]

**Pull request merged**
